src/c++/perf_analyzer/test_custom_load_manager.cc

```diff
@@ -0,0 +1,117 @@
+
+#include <chrono>
+#include <memory>
+#include <string>
+#include <vector>
+#include "client_backend/client_backend.h"
+#include "constants.h"
+#include "custom_load_manager.h"
+#include "doctest.h"
+#include "request_rate_manager.h"
+
+namespace triton { namespace perfanalyzer {
+
+class TestCustomLoadManager : public CustomLoadManager {
+ public:
+  std::unique_ptr<std::chrono::nanoseconds>& gen_duration_{
+      RequestRateManager::gen_duration_};
+  std::vector<std::chrono::nanoseconds>& schedule_{
+      RequestRateManager::schedule_};
+  std::string& request_intervals_file_{
+      CustomLoadManager::request_intervals_file_};
+  std::vector<std::chrono::nanoseconds>& custom_intervals_{
+      CustomLoadManager::custom_intervals_};
+
+  cb::Error ReadTimeIntervalsFile(
+      const std::string& path,
+      std::vector<std::chrono::nanoseconds>* contents) override
+  {
+    return cb::Error::Success;
+  }
+};
+
+TEST_CASE("testing the InitCustomIntervals function")
+{
+  TestCustomLoadManager tclm{};
+
+  SUBCASE("no file provided")
+  {
+    cb::Error result{tclm.InitCustomIntervals()};
+
+    CHECK(result.Err() == SUCCESS);
+    CHECK(tclm.schedule_.size() == 1);
+    CHECK(tclm.schedule_[0] == std::chrono::nanoseconds(0));
+  }
+
+  SUBCASE("file provided")
+  {
+    tclm.request_intervals_file_ = "nonexistent_file.txt";
+    tclm.gen_duration_ = std::make_unique<std::chrono::nanoseconds>(350000000);
+    tclm.custom_intervals_.push_back(std::chrono::nanoseconds(100000000));
+    tclm.custom_intervals_.push_back(std::chrono::nanoseconds(110000000));
+    tclm.custom_intervals_.push_back(std::chrono::nanoseconds(130000000));
+
+    cb::Error result{tclm.InitCustomIntervals()};
+
+    CHECK(result.Err() == SUCCESS);
+    CHECK(tclm.schedule_.size() == 5);
+    CHECK(tclm.schedule_[0] == std::chrono::nanoseconds(0));
+    CHECK(tclm.schedule_[1] == std::chrono::nanoseconds(100000000));
+    CHECK(tclm.schedule_[2] == std::chrono::nanoseconds(210000000));
+    CHECK(tclm.schedule_[3] == std::chrono::nanoseconds(340000000));
+    CHECK(tclm.schedule_[4] == std::chrono::nanoseconds(440000000));
+  }
+}
+
+TEST_CASE("testing the GetCustomRequestRate function")
+{
+  TestCustomLoadManager tclm{};
+  double request_rate{0.0};
+
+  SUBCASE("custom_intervals_ empty")
+  {
+    cb::Error result{tclm.GetCustomRequestRate(&request_rate)};
+
+    CHECK(result.Err() == GENERIC_ERROR);
+    CHECK(result.Message() == "The custom intervals vector is empty");
+  }
+
+  SUBCASE("custom_intervals_ populated")
+  {
+    tclm.custom_intervals_.push_back(std::chrono::nanoseconds(100000000));
+    tclm.custom_intervals_.push_back(std::chrono::nanoseconds(110000000));
+    tclm.custom_intervals_.push_back(std::chrono::nanoseconds(130000000));
+
+    cb::Error result{tclm.GetCustomRequestRate(&request_rate)};
+
+    CHECK(result.Err() == SUCCESS);
+    CHECK(request_rate == doctest::Approx(8.0));
```

Comment:
`GetCustomRequestRate` basically just does this:

floor(num_custom_intervals * 1 billion / custom intervals sum)

Reply:
While I agree with this testing of the current behavior, I'm also left wondering if this is correct. In the same follow-on bug I'd like to see some more tests that confirm the results are what they *should* be, not what the current code is doing.

Reply:
> I'm also left wondering if this is correct.

It seems like a design choice to try to present the user with some "representative" request rate based on the custom intervals. I think it's fine to take an "average" as it is calculated. What other representative request rate would lead to a good UX, in your opinion?

Reply:
> I'd like to see some more tests that confirm the results are what they should be

Can you give an example? There are already tests that test the scheduling code for request rate, so it seems only pertinent for the custom load manager tests to test methods directly related to it (defined in it).

Reply:
This is probably fine once the issue in my other comment is resolved, but something to think about for that case.